Draft: Apni2/tmc5xxx log status

apni2 opened the pull request:
Move SPI status byte logging to tmc50xx and tmc51xx.
Add decoding option controlled by STEPPER_ADI_TMC50XX_LOG_STATUS and STEPPER_ADI_TMC51XX_LOG_STATUS.

Should be combined with: https://github.com/zephyrproject-rtos/zephyr/pull/88795

faxe1008 reviewed (changes requested):
I get the idea that it's for diagnostics, but its unfortunate that also users which are not interested in this have to pay the price in flash and compute. A DBG_LOG of the raw status register would be good enough here. If you need to diagnose any kinds of issue you are using the datasheet anyways (hopefully), so translation of the bit to the code is an excersise for the reader.

jilaypandya commented:
> I get the idea that it's for diagnostics, but its unfortunate that also users which are not interested in this have to pay the price in flash and compute. A DBG_LOG of the raw status register would be good enough here. If you need to diagnose any kinds of issue you are using the datasheet anyways (hopefully), so translation of the bit to the code is an excersise for the reader.

apni2 commented on drivers/stepper/adi_tmc/adi_tmc_spi.c:
```diff
@@ -13,6 +14,44 @@
+	LOG_DBG("%s", buf);
+}
+
 static void parse_tmc_spi_status(const uint8_t status_byte)
```
I am not entirely sure how to share code between a 50xx and 51xx implementation when both controllers must be supported in the same build. But I guess a tmc_spi could call a specific decode function or the driver api could return the status byte.

apni2 replied:
For an updated version supporting both tmx50xx and tmc51xx see: https://github.com/apni2/zephyr/commit/c6917014bc8eb4bc71f9fe5a22aa715893a63ae1

apni2 commented:
> I get the idea that it's for diagnostics, but its unfortunate that also users which are not interested in this have to pay the price in flash and compute. A DBG_LOG of the raw status register would be good enough here. If you need to diagnose any kinds of issue you are using the datasheet anyways (hopefully), so translation of the bit to the code is an excersise for the reader.

I agree and definitely think it is a good idea to have this encoding controlled by a KConfig option. However I did find it to be helpful in order to understand how the controller operates when running a motor. Perhaps this PL should just be paused for now?

jilaypandya commented:
I think it's best to put these kind of board-bring up logs directly within the driver code, ofcourse guarded by a kconfig variable. Thanks for this PR, i need to do some housekeeping with the tmc50xx driver :)

fabiobaltieri commented on drivers/stepper/adi_tmc/adi_tmc_spi.c:
```diff
@@ -13,6 +14,44 @@
+static void log_status(uint8_t status_byte)
+{
+	static const char *spi_status[8] = {
+#if defined CONFIG_STEPPER_ADI_TMC50XX
```
`#ifdef`

fabiobaltieri commented on drivers/stepper/adi_tmc/adi_tmc_spi.c:
```diff
@@ -13,6 +14,44 @@
+#endif
+	};
+	char buf[110];
+	int n = sprintf(buf, "0x%02x", status_byte);
```
`snprintf`, just as good practice, code is copy pasted around better be defensive

apni2 replied:
Done

apni2 replied:
Updated.

jilaypandya commented:
> I agree and definitely think it is a good idea to have this encoding controlled by a KConfig option. However I did find it to be helpful in order to understand how the controller operates when running a motor. Perhaps this PL should just be closed or paused for now?
 
I think this status byte logging has to be made part of the tmc50xx_stepper_controller.c, for which there is already an ongoing pr. We can create an extra kconfig template for logging stallguard & status byte values and and handle it on per driver basis.

apni2 commented:
I ran into issues testing the tmc5160 driver so updated this PR. First version of tmc5160 support is done and I will be creating a PR for that shortly.

apni2 commented:
I ran into issues testing the tmc5160 driver so updated this PR. First version of tmc5160 support is done and I will be creating a PR for that shortly.

apni2 commented:
For an updated version supporting both tmx50xx and tmc51xx see: https://github.com/apni2/zephyr/commit/c6917014bc8eb4bc71f9fe5a22aa715893a63ae1

apni2 commented:
Is this PL blocked or can it be merged?

jilaypandya commented:
> Is this PL blocked or can it be merged?

For a PR to get merged you need approval from the Assignees, which in this case is me, one further approval from a member of the project and the requested changes have to be resolved

jilaypandya commented:
Hi @apni2, can you have a look at this PR https://github.com/zephyrproject-rtos/zephyr/pull/88795? Parsing of status byte is something that we have decided not to do in upstream code.

apni2 committed:
drivers: stepper: adi_tmc: log SPI status

Move SPI status byte logging to tmc50xx and tmc51xx.
Add decoding option controlled by: STEPPER_ADI_TMC51XX_LOG_STATUS and
STEPPER_ADI_TMC51XX_LOG_STATUS.
Log GCONF as hex value. Include log options in build_all prj.conf.

Signed-off-by: Anders Nielsen <anders.nielsen@prevas.dk>

jilaypandya commented:
> > Hi @apni2, can you have a look at this PR #88795? Parsing of status byte is something that we have decided not to do in upstream code.
> 
> PR #88795 looks fine. I don't know how to approve (or if I am able to do that), but left a comment.
> 
> This PR has been update to do status byte decoding in the driver specific code and includes the changes from #88795. GCONF is now logged in hex format and log options are added to build_all prj.conf.
> 
> Would you consider this PR upstream if SPI status log is off by default via the Kconfig option?

apni2 commented:
> What i meant is just use the print_rx_tx_buffer as rx_buffer[0] is status byte and drop parsing of status byte using callback as you are doing now.
> 
> I kind of proposed the same thing in #88795 and we decided not to go ahead with it :)

I see your point and agree that this PR does not add much without the decoding part so I will close it.
